internal/consensus/state.go

```diff
@@ -2544,6 +2572,63 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
 	return added, err
 }
 
+func (cs *State) AddCommit(commit *types.Commit, peerID p2p.ID) (added bool, err error) {
+	cs.Logger.Debug(
+		"Adding whole commit",
+		"commit_height", commit.Height,
+		"commit_round", commit.Round,
+		"commit_blockId", commit.BlockID,
+		"cs_height", cs.Height,
+		"cs_round", cs.Round,
+	)
+
+	if commit.Height < cs.Height {
+		// cs.metrics.MarkLateCommit() TODO Implement
+	}
+
+	// Height mismatch is ignored.
+	// Not necessarily a bad peer, but not favorable behavior.
+	if commit.Height != cs.Height {
+		cs.Logger.Debug("Commit ignored and not added",
+			"commit_height", commit.Height,
+			"cs_height", cs.Height,
+			"peer", peerID)
+		return added, err
+	}
+
+	// Check to see if the chain is configured to extend votes.
+	extEnabled := cs.isVoteExtensionsEnabled(commit.Height)
+	if extEnabled {
+		// We don't support receiving commits with vote extensions enabled ATM.
+		cs.Logger.Error("Received commit with vote extension for height %v (extensions disabled) from peer ID %s", commit.Height, peerID)
+		return added, err
+	}
+
+	// TODO: double check that we have fully validated the commit at this function's caller
+
+	height := cs.Height
+	cs.Votes.SetCommit(commit)
+
+	// TODO We need something similar for Commits
+	// if err := cs.eventBus.PublishEventVote(types.EventDataVote{Vote: vote}); err != nil {
+	// 	return added, err
+	// }
+	// cs.evsw.FireEvent(types.EventVote, vote)
+
+	// Executed as Commit could be from a higher round
+	cs.enterNewRound(height, commit.Round)
+	cs.enterPrecommit(height, commit.Round)
+
+	cs.enterCommit(height, commit.Round)
+	// TODO Enable back. Does Commit have something similar? If not implement HasAll() for Commit
+	// skipTimeoutCommit := cs.state.NextBlockDelay == 0 && cs.config.TimeoutCommit == 0
+	// if skipTimeoutCommit && commit.HasAll() {
```

Comment:
This function should only be called when a node is in catch-up mode. This there is not much point in using timeout commit. We should delete this. Or immediately call `enterNewRound`